src/test/shell/bazel/bazel_cc_code_coverage_test.sh

```diff
@@ -0,0 +1,134 @@
+#!/bin/bash -eu
+#
+# Copyright 2018 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Unit tests for tools/test/collect_cc_code_coverage.sh
+
+
+# Load the test setup defined in the parent directory
+CURRENT_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+source "${CURRENT_DIR}/../integration_test_setup.sh" \
+  || { echo "integration_test_setup.sh not found!" >&2; exit 1; }
+
+
+function setup_cc_sources() {
+  mkdir coverage
+  cat << EOF > coverage/a.h
+int a(bool what);
+EOF
+
+  cat << EOF > coverage/a.cc
+#include "a.h"
+
+int a(bool what) {
+  if (what) {
+    return 1;
+  } else {
+    return 2;
+  }
+}
+EOF
+
+  cat << EOF > coverage/t.cc
+#include <stdio.h>
+#include "a.h"
+
+int main(void) {
+  a(true);
+}
+EOF
+}
+
+function setup_gcc_gcda_files() {
+  cd coverage/
+  # Generate .gcno files.
+  g++ -fprofile-arcs -ftest-coverage -c a.h a.cc t.cc
+  # Produce instrumented binary.
+  g++ -fprofile-arcs -ftest-coverage a.h a.cc t.cc -o test
+  # Generate test.gcda file.
+  ./test
+  cd ..
+}
+
+function setup_script_environment() {
+  export COVERAGE_DIR="$PWD/coverage"
+  export ROOT="$PWD"
+  export COVERAGE_OUTPUT_FILE="$PWD/coverage_report.dat"
+  export COVERAGE_MANIFEST="$PWD/coverage_manifest.txt"
+
+  # The script expects gcov to be at $COVERAGE_DIR/gcov.
+  gcov_location=$( which gcov )
+  cp $gcov_location $COVERAGE_DIR/gcov
+
+  # The script expects the output file to already exist.
+  touch $COVERAGE_OUTPUT_FILE
+  echo "coverage/a.gcno" >> $COVERAGE_MANIFEST
+}
+
+function check_env() {
+  if [[ ! -x /usr/bin/lcov ]]; then
+    fail "lcov not installed. Skipping test."
+  fi
+
+  if [[ -z $( which gcov ) ]]; then
+    fail "gcov not installed."
+  fi
+
+  if [[ -z $( which g++ ) ]]; then
+    fail "g++ not installed."
+  fi
+}
+
+function test_cc_test_coverage() {
+  check_env
```

Comment:
All of these "setup" and "check" should be part of the "set_up" function of the unittest framework we use for these tests, and they should be generic. Same goes for steps after that can go in the tear_down function.

Reply:
This is very specific for this test and one important difference is that the test does not invoke bazel. Having these setups in the unittest framework could interfere with other tests environment.

Reply:
Oh sorry, that's not what I meant :-). Since you're loading the integration_test_setup then you're loading unittest.bash which has a nice functionality where you can define a set_up and a tear_down method that runs before and after each test.